be/src/storage/rowset/beta_rowset_writer.cpp

```diff
@@ -53,13 +53,121 @@ BetaRowsetWriter::BetaRowsetWriter(const RowsetWriterContext& context)
+             _context.tablet_schema->contains_format_v2_column()));
+    auto real_data_format = _context.storage_format_version;
+    auto tablet_format = real_data_format;
+    if (_context.tablet_schema->contains_format_v1_column()) {
+        tablet_format = kDataFormatV1;
+    } else if (_context.tablet_schema->contains_format_v2_column()) {
+        tablet_format = kDataFormatV2;
+    }
+
+    // StarRocks is built on earlier work on Apache Doris and is compatible with its data format but
+    // has newly designed storage formats for DATA/DATETIME/DECIMAL for better performance.
+    // When loading data into a tablet created by Apache Doris, the real data format will
+    // be different from the tablet schema, so here we create a new schema matched with the
+    // real data format to init `SegmentWriter`.
+    if (real_data_format != tablet_format) {
+        _rowset_schema = _context.tablet_schema->convert_to_format(real_data_format);
+    }
+
+    _rowset_meta = std::make_shared<RowsetMeta>();
+    _rowset_meta->set_rowset_id(_context.rowset_id);
+    _rowset_meta->set_partition_id(_context.partition_id);
+    _rowset_meta->set_tablet_id(_context.tablet_id);
+    _rowset_meta->set_tablet_schema_hash(_context.tablet_schema_hash);
+    _rowset_meta->set_rowset_type(_context.rowset_type);
+    _rowset_meta->set_rowset_state(_context.rowset_state);
+    _rowset_meta->set_segments_overlap(_context.segments_overlap);
+    if (_context.rowset_state == PREPARED || _context.rowset_state == COMMITTED) {
+        _is_pending = true;
+        _rowset_meta->set_txn_id(_context.txn_id);
+        _rowset_meta->set_load_id(_context.load_id);
+    } else {
+        _rowset_meta->set_version(_context.version);
+    }
+    _rowset_meta->set_tablet_uid(_context.tablet_uid);
+    return OLAP_SUCCESS;
+}
+
+RowsetSharedPtr BetaRowsetWriter::build() {
+    _rowset_meta->set_num_rows(_num_rows_written);
+    _rowset_meta->set_total_row_size(_total_row_size);
+    _rowset_meta->set_total_disk_size(_total_data_size);
+    _rowset_meta->set_data_disk_size(_total_data_size);
+    _rowset_meta->set_index_disk_size(_total_index_size);
+    // TODO write zonemap to meta
+    _rowset_meta->set_empty(_num_rows_written == 0);
+    _rowset_meta->set_creation_time(time(nullptr));
+    _rowset_meta->set_num_segments(_num_segment);
+    // newly created rowset do not have rowset_id yet, use 0 instead
+    _rowset_meta->set_rowset_seg_id(0);
+    // updatable tablet require extra processing
+    if (_context.tablet_schema->keys_type() == KeysType::PRIMARY_KEYS) {
+        if (_segment_has_deletes.size() >= 1 && _segment_has_deletes[0]) {
+            _rowset_meta->set_num_delete_files(1);
+        } else {
+            _rowset_meta->set_num_delete_files(0);
+        }
+        _rowset_meta->set_segments_overlap(NONOVERLAPPING);
+    } else {
+        if (_num_segment <= 1) {
+            _rowset_meta->set_segments_overlap(NONOVERLAPPING);
+        }
```

Comment:
else?

Reply:
Segments overlap is set according to the context when writer inits.
This is the special case for non overlapping.